x-pack/solutions/observability/plugins/apm/server/routes/service_map/fetch_service_paths_from_trace_ids.ts

```diff
@@ -5,345 +5,319 @@
  * 2.0.
  */
 
-import { rangeQuery } from '@kbn/observability-plugin/server';
-import { ProcessorEvent } from '@kbn/observability-plugin/common';
+import { rangeQuery, termsQuery } from '@kbn/observability-plugin/server';
+import type { QueryDslQueryContainer } from '@elastic/elasticsearch/lib/api/types';
 import {
   AGENT_NAME,
   PARENT_ID,
-  PROCESSOR_EVENT,
   SERVICE_ENVIRONMENT,
   SERVICE_NAME,
   SPAN_DESTINATION_SERVICE_RESOURCE,
+  SPAN_ID,
   SPAN_SUBTYPE,
   SPAN_TYPE,
+  TIMESTAMP_US,
   TRACE_ID,
+  TRANSACTION_ID,
 } from '../../../common/es_fields/apm';
 import type {
   ConnectionNode,
+  DiscoveredService,
   ExternalConnectionNode,
   ServiceConnectionNode,
 } from '../../../common/service_map';
-import type { APMEventClient } from '../../lib/helpers/create_es_client/create_apm_event_client';
-import { calculateDocsPerShard } from './calculate_docs_per_shard';
+import type { EsClient } from '../../lib/helpers/get_esql_client';
 
-const SCRIPTED_METRICS_FIELDS_TO_COPY = [
-  PARENT_ID,
-  SERVICE_NAME,
-  SERVICE_ENVIRONMENT,
-  SPAN_DESTINATION_SERVICE_RESOURCE,
-  TRACE_ID,
-  PROCESSOR_EVENT,
-  SPAN_TYPE,
-  SPAN_SUBTYPE,
-  AGENT_NAME,
-];
+type QueryResult = {
+  'event.id': string;
+  [PARENT_ID]?: string;
+} & ConnectionNode;
 
-const AVG_BYTES_PER_FIELD = 55;
+type Node = {
+  id: string;
+  children: Node[];
+} & ConnectionNode;
 
 export async function fetchServicePathsFromTraceIds({
-  apmEventClient,
   traceIds,
   start,
   end,
+  index,
+  filters,
+  esqlClient,
   terminateAfter,
-  serviceMapMaxAllowableBytes,
-  numOfRequests,
 }: {
-  apmEventClient: APMEventClient;
   traceIds: string[];
   start: number;
   end: number;
   terminateAfter: number;
-  serviceMapMaxAllowableBytes: number;
-  numOfRequests: number;
+  esqlClient: EsClient;
+  index: string[];
+  filters: QueryDslQueryContainer[];
 }) {
-  // make sure there's a range so ES can skip shards
-  const dayInMs = 24 * 60 * 60 * 1000;
-  const startRange = start - dayInMs;
-  const endRange = end + dayInMs;
-
-  const serviceMapParams = {
-    apm: {
-      events: [ProcessorEvent.span, ProcessorEvent.transaction],
-    },
-    body: {
-      terminate_after: terminateAfter,
-      track_total_hits: false,
-      size: 0,
-      query: {
+  // The query does the following:
+  // 1. Gets the latest field values for each event.id
+  //  Due to the  lack of esql native aggregation to get the latest value of a field,
```

Comment:
Is this something we plan to look into? Do we know if this would substantially improve performance?

Reply:
Yes. @nik9000 and @alex-spies can give more details on possible performance gains, but the idea is to have a `TOP` aggregation or a `LIMIT 1 BY` syntax to replace that.